File: rtt.cpp
#include <iostream>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <chrono>

int main() {
    const int ITERATIONS = 100'000;
    
    std::mutex mtx;
    std::condition_variable cv;
    // 0 = A's turn, 1 = B's turn, -1 = Stop
    int turn = 0; 

    // Thread B (The Responder)
    std::thread thread_b([&]() {
        while (true) {
            std::unique_lock<std::mutex> lock(mtx);
            // Wait while it is A's turn
            cv.wait(lock, [&]{ return turn != 0; });
            
            if (turn == -1) break; // Exit condition
            
            turn = 0; // Give turn back to A
            lock.unlock(); // Drop lock before notifying to reduce RTT overhead
            cv.notify_one();
        }
    });

    // Thread A (The Measurer)
    auto total_rtt = std::chrono::nanoseconds(0);

    for (int i = 0; i < ITERATIONS; ++i) {
        auto start = std::chrono::high_resolution_clock::now();
        
        {
            std::lock_guard<std::mutex> lock(mtx);
            turn = 1; // Give turn to B
        }
        cv.notify_one();
        
        {
            std::unique_lock<std::mutex> lock(mtx);
            // Wait while it is B's turn
            cv.wait(lock, [&]{ return turn == 0; });
        }
        
        auto end = std::chrono::high_resolution_clock::now();
        total_rtt += std::chrono::duration_cast<std::chrono::nanoseconds>(end - start);
    }

    // Stop Thread B
    {
        std::lock_guard<std::mutex> lock(mtx);
        turn = -1;
    }
    cv.notify_one();
    thread_b.join();

    std::cout << "Average C++ CV RTT: " 
              << total_rtt.count() / ITERATIONS << " ns\n";

    return 0;
}

File: rtt.rs
use std::sync::{Arc, Mutex, Condvar};
use std::thread;
use std::time::{Instant, Duration};

fn main() {
    const ITERATIONS: u32 = 100_000;

    // 0 = A's turn, 1 = B's turn, -1 = Stop
    let state = Arc::new((Mutex::new(0), Condvar::new()));
    let state_clone = Arc::clone(&state);

    // Thread B (The Responder)
    let responder = thread::spawn(move || {
        let (lock, cvar) = &*state_clone;
        
        loop {
            let mut turn = lock.lock().unwrap();
            // Wait while it is A's turn (0)
            turn = cvar.wait_while(turn, |t| *t == 0).unwrap();
            
            if *turn == -1 { break; } // Exit condition
            
            *turn = 0; // Give turn back to A
            drop(turn); // Drop lock before notifying
            cvar.notify_one();
        }
    });

    // Thread A (The Measurer)
    let (lock, cvar) = &*state;
    let mut total_rtt = Duration::new(0, 0);

    for _ in 0..ITERATIONS {
        let start = Instant::now();
        
        {
            let mut turn = lock.lock().unwrap();
            *turn = 1; // Give turn to B
        }
        cvar.notify_one();

        {
            // FIX 1: Removed 'mut' because we only read the value here
            let turn = lock.lock().unwrap(); 
            
            // FIX 2: Used drop() instead of 'let _ =' to explicitly release the lock
            drop(cvar.wait_while(turn, |t| *t != 0).unwrap()); 
        }
        
        total_rtt += start.elapsed();
    }

    // Stop Thread B
    {
        let mut turn = lock.lock().unwrap();
        *turn = -1;
    }
    cvar.notify_one();
    responder.join().unwrap();

    println!("Average Rust CV RTT: {} ns", total_rtt.as_nanos() / ITERATIONS as u128);
}
